pkg/clusteragent/autoscaling/workload/controller.go

```diff
@@ -0,0 +1,333 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver
+
+package workload
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	scaleclient "k8s.io/client-go/scale"
+	"k8s.io/utils/clock"
+
+	datadoghq "github.com/DataDog/datadog-operator/apis/datadoghq/v1alpha1"
+
+	"github.com/DataDog/datadog-agent/pkg/clusteragent/autoscaling"
+	"github.com/DataDog/datadog-agent/pkg/clusteragent/autoscaling/workload/model"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+const (
+	maxRetry int = 5
+
+	controllerID = "dpa-c"
+)
+
+var (
+	podAutoscalerGVR  = datadoghq.GroupVersion.WithResource("datadogpodautoscalers")
+	podAutoscalerMeta = metav1.TypeMeta{
+		Kind:       "DatadogPodAutoscaler",
+		APIVersion: "datadoghq.com/v1alpha1",
+	}
+)
+
+type store = autoscaling.Store[model.PodAutoscalerInternal]
+
+type processResult struct {
+	autoscaling.ProcessResult
+	updateStatus bool
+}
+
+func withStatusUpdate(updateStatus bool, result autoscaling.ProcessResult) processResult {
+	return processResult{
+		ProcessResult: result,
+		updateStatus:  updateStatus,
+	}
+}
+
+// Controller for DatadogPodAutoscaler objects
+type Controller struct {
+	*autoscaling.Controller
+
+	clock clock.Clock
+	store *store
+
+	horizontalController *horizontalController
+}
+
+// newController returns a new workload autoscaling controller
+func newController(
+	restMapper apimeta.RESTMapper,
+	scaleClient scaleclient.ScalesGetter,
+	dynamicClient dynamic.Interface,
+	dynamicInformer dynamicinformer.DynamicSharedInformerFactory,
+	isLeader func() bool,
+	store *store,
+) (*Controller, error) {
+	c := &Controller{
+		clock: clock.RealClock{},
+	}
+
+	baseController, err := autoscaling.NewController(controllerID, c, dynamicClient, dynamicInformer, podAutoscalerGVR, isLeader, store)
+	if err != nil {
+		return nil, err
```

Comment:
If the informer fails to start, I suppose there is no retry mechanism and the cluster agent will stay healthy. Should we add one or consider adding one in the future ?

Reply:
After double-checking, it appears only `AddEventHandler` can fail and it fails only if the informer is stopped so we're good